src/err/err_state.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{
+    exceptions::PyBaseException, ffi, types::PyType, IntoPy, IntoPyPointer, Py, PyObject, Python,
+};
+
+#[derive(Clone)]
+pub(crate) struct PyErrStateNormalized {
+    pub ptype: Py<PyType>,
+    pub pvalue: Py<PyBaseException>,
+    pub ptraceback: Option<PyObject>,
+}
+
+pub(crate) enum PyErrState {
+    Lazy {
+        ptype: Py<PyType>,
+        pvalue: Box<dyn FnOnce(Python) -> PyObject + Send + Sync>,
+    },
+    FfiTuple {
+        ptype: Option<PyObject>,
+        pvalue: Option<PyObject>,
+        ptraceback: Option<PyObject>,
+    },
+    Normalized(PyErrStateNormalized),
+}
+
+/// Helper conversion trait that allows to use custom arguments for lazy exception construction.
+pub trait PyErrArguments: Send + Sync {
+    /// Arguments for exception
+    fn arguments(self, py: Python) -> PyObject;
+}
+
+impl<T> PyErrArguments for T
+where
+    T: IntoPy<PyObject> + Send + Sync,
+{
+    fn arguments(self, py: Python) -> PyObject {
+        self.into_py(py)
+    }
+}
+
+pub(crate) fn boxed_args(
+    args: impl PyErrArguments + 'static,
+) -> Box<dyn FnOnce(Python) -> PyObject + Send + Sync> {
+    Box::new(|py| args.arguments(py))
+}
+
+impl PyErrState {
+    pub fn into_ffi_tuple(
```

Comment:
`PyErrState` is `pub (crate)` but this fn is `pub`, given that this function returns raw pointers, I'd expect that this is not used externally. Could this also be marked `pub (crate)`? Otherwise, adding some documentation would be nice!